File: 0120-三角形最小路径和/Solution.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        if(triangle == null || triangle.size() == 0 || triangle.get(0).size() == 0){
            return 0;
        }
        int length = triangle.get(triangle.size()-1).size();
        int[][] dp = new int[length][length];
        dp[0][0] = triangle.get(0).get(0);
        for (int i = 1; i < length; i++) {
            for (int j = 0; j <= i; j++) {
                dp[i][j] = triangle.get(i).get(j);
                if(j == 0){
                    dp[i][j] += dp[i-1][0];
                }else if(j == i){
                    dp[i][j] += dp[i-1][j-1];
                }else{
                    dp[i][j] += Math.min(dp[i-1][j], dp[i-1][j-1]);
                }
            }
        }
        int res = Integer.MAX_VALUE;
        for(int i=0; i< length; i++){
            if(dp[length-1][i] < res){
                res = dp[length-1][i];
            }
        }
        return res;
    }
}

File: 0120-三角形最小路径和/Solution.py
class Solution:
    def __init__(self):
        self.cache = {}

    def minimumTotal(self, triangle: List[List[int]]) -> int:
        if not triangle:
            return 0
        ret = self.recursion(triangle, 0, 0)
        return ret

    def recursion(self, triangle, i, j):
        if i >= len(triangle):
            return 0
        if (i, j) in self.cache:
            return self.cache[(i, j)]
        curr = triangle[i][j]
        left = self.recursion(triangle, i+1, j)
        right = self. recursion(triangle, i+1, j+1)
        curr += min(left, right)
        self.cache[(i, j)] = curr
        return curr

    # 迭代
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        if not triangle:
            return 0
        table = [[0] * len(line) for line in triangle]
        table[0][0] = triangle[0][0]
        for i in range(1, len(table)):
            for j in range(1, len(table[i])-1):
                table[i][j] = triangle[i][j] + min(table[i-1][j-1], table[i-1][j])
            table[i][0] = triangle[i][0] + table[i-1][0]
            table[i][-1] = triangle[i][-1] + table[i-1][-1]
        return min(table[-1])